KAFKA-8730: Add API to delete consumer offsets (KIP-496)

dajac opened the pull request:
This adds an administrative API to delete consumer offsets of a group as well as extends the mechanism to expire offsets of consumer groups.

It makes the group coordinator aware of the set of topics a consumer group (protocol type == 'consumer') is actively subscribed to, allowing offsets of topics which are not actively subscribed to by the group to be either expired or administratively deleted. The expiration rules remain the same.

For the other groups (non-consumer), the API allows to delete offsets when the group is empty and the expiration remains the same.

*More detailed description of your change,
if necessary. The PR title and PR message become
the squashed commit message, so use a separate
comment to ping reviewers.*

*Summary of testing strategy (including rationale)
for the feature or bug fix. Unit and/or integration
tests are expected for any behaviour change and
system tests should be considered for larger changes.*

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

dajac committed:
KAFKA-8730: Add API to delete consumer offsets (KIP-496)

stanislavkozlovski commented on clients/src/main/java/org/apache/kafka/common/errors/GroupSubscribedToTopicException.java:
```diff
@@ -0,0 +1,23 @@
+ */
+package org.apache.kafka.common.errors;
+
+public class GroupSubscribedToTopicException extends ApiException {
```
This is a public interface and wasn't mentioned in the KIP. Was there an offline discussion on adding this?

dajac replied:
Right. There was not an offline discussion about it. I have forgotten to bring the point. What's the best way? Shall I discuss thing in the voting thread?

To give the context: When one tried to delete the offsets of an active consumer group, we need a way to indicate if the topic is still actively consumed by the group which means that it is not possible to delete the offsets.

dajac replied:
@hachikuji What do you think about adding this one?

gwenshap replied:
I think this makes sense, and you can bring it up in the vote thread and ask if anyone objects.

gwenshap replied:
Good catch @stanislavkozlovski :)

hachikuji replied:
Seems reasonable to me.

gwenshap commented:
Is there a reason we are not using the protocol generation for the new protocol?

dajac commented:
> Is there a reason we are not using the protocol generation for the new protocol?

it is used. See the json files which generates OffsetDeleteRequestData and OffsetDeleteResponseData which are encapsulated respectively in OffsetDeleteRequest and OffsetDeleteResponse. I have followed what has been done for other requests/responses: e.g. https://github.com/apache/kafka/blob/trunk/clients/src/main/java/org/apache/kafka/common/requests/ElectLeadersRequest.java

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/admin/DeleteConsumerGroupOffsetsResult.java:
```diff
@@ -0,0 +1,85 @@
+                if (throwable != null) {
+                    result.completeExceptionally(throwable);
+                } else if (!topicPartitions.containsKey(partition)) {
+                    result.completeExceptionally(new UnknownTopicOrPartitionException(
```
This handles the case of a topic that wasn't in the requested set of partitions? Seems like `IllegalArgumentException` might be more appropriate?

dajac replied:
You're right. I say `UnknownTopicOrPartitionException` here [1] so I thought it was the way to go in the Admin client. `IllegalArgumentException` is better.

[1] https://github.com/apache/kafka/blob/trunk/clients/src/main/java/org/apache/kafka/clients/admin/ElectPreferredLeadersResult.java#L62

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/ConsumerProtocol.java:
```diff
@@ -120,6 +120,11 @@
         new Field(TOPIC_PARTITIONS_KEY_NAME, new ArrayOf(TOPIC_ASSIGNMENT_V0)),
         new Field(USER_DATA_KEY_NAME, Type.NULLABLE_BYTES));
 
+    public static Short deserializeHeader(ByteBuffer buffer) {
```
nit: maybe `deserializeVersion` so that the result is part of the name?

dajac replied:
I agree. I will change it.

hachikuji commented on clients/src/main/java/org/apache/kafka/common/protocol/Errors.java:
```diff
@@ -312,7 +313,9 @@
     NO_REASSIGNMENT_IN_PROGRESS(85, "No partition reassignment is in progress.",
-            NoReassignmentInProgressException::new);
+            NoReassignmentInProgressException::new),
+    GROUP_SUBSCRIBED_TO_TOPIC(86, "The consumer group is actively subscribed to the topic",
```
Might be worth adding a little more detail about the context for this error.

dajac replied:
Ack. I will make it better.

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java:
```diff
@@ -3053,6 +3060,98 @@ void handleFailure(Throwable throwable) {
+                    });
+                });
+
+                // If the error is an error at the group level, the future is failed with it
```
Some other group level APIs have a top-level error code (e.g. OffsetFetchResponse). Do you think it's worth adding one?

dajac replied:
Yeah. It is worth adding one. It will avoid having to handle global errors manually in the admin client. Also, it feels cleaner.

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java:
```diff
@@ -3053,6 +3060,98 @@ void handleFailure(Throwable throwable) {
+                    final Errors error = partitions.values().iterator().next();
+                    if (error == Errors.COORDINATOR_LOAD_IN_PROGRESS || error == Errors.COORDINATOR_NOT_AVAILABLE) {
+                        throw error.exception();
+                    } else if (error == Errors.INVALID_GROUP_ID || error == Errors.GROUP_ID_NOT_FOUND) {
```
Another one is `GROUP_AUTHORIZATION_FAILED`.

hachikuji commented on clients/src/main/java/org/apache/kafka/common/requests/OffsetDeleteRequest.java:
```diff
@@ -0,0 +1,113 @@
+    }
+
+    public AbstractResponse getErrorResponse(int throttleTimeMs, Errors error) {
+        switch (version()) {
```
Hmm. I think we're just following the same pattern used elsewhere, but I think the version check is redundant. We already know it's a valid version because the request constructor verifies it.

dajac replied:
I have checked and your are right. I have removed it.

hachikuji commented on clients/src/main/java/org/apache/kafka/common/requests/OffsetDeleteResponse.java:
```diff
@@ -0,0 +1,89 @@
+/**
+ * Possible error codes:
+ * <p>
+ * COORDINATOR_LOAD_IN_PROGRESS (14)
```
nit: Can we link to the error code as we do in `OffsetFetchResponse`?

dajac replied:
Sure thing.

hachikuji commented on clients/src/main/java/org/apache/kafka/common/requests/OffsetDeleteResponse.java:
```diff
@@ -0,0 +1,89 @@
+
+    @Override
+    public boolean shouldClientThrottle(short version) {
+        return version >= 1;
```
Shouldn't this be 0?

dajac replied:
Indeed. Good catch.

hachikuji commented on clients/src/test/java/org/apache/kafka/clients/admin/KafkaAdminClientTest.java:
```diff
@@ -1517,6 +1526,152 @@ public void testDeleteConsumerGroups() throws Exception {
+        try (AdminClientUnitTestEnv env = new AdminClientUnitTestEnv(cluster)) {
+            env.kafkaClient().setNodeApiVersions(NodeApiVersions.create());
+
+            //Retriable FindCoordinatorResponse errors should be retried
```
Any chance we could break this test up? I think each of these comments could probably be a separate test case.

dajac replied:
Yes. I have should have done this before.

hachikuji reviewed (commented):
Thanks for the patch. I made a pass on the client logic. I will look at the broker implementation next.

dajac committed:
Address reviews

dajac committed:
Address reviews

dajac commented:
@hachikuji I have addressed all your comments. I have refactored the response to include a top level error. It is cleaner like this. We need to bring the change in the KIP. I will do it if you are fine with the changes.

hachikuji commented:
@dajac Sounds good to me. I'll try to review the rest today.

hachikuji commented on core/src/main/scala/kafka/coordinator/group/GroupMetadataManager.scala:
```diff
@@ -95,6 +96,26 @@ class GroupMetadataManager(brokerId: Int,
+  val offsetCommitsSensor = metrics.sensor("OffsetCommits")
+
+  offsetCommitsSensor.add(new Meter(
+    metrics.metricName("offset-commits-rate",
```
nit: maybe `offset-commit-rate`?

hachikuji commented on core/src/main/scala/kafka/coordinator/group/GroupMetadata.scala:
```diff
@@ -423,6 +434,51 @@ private[group] class GroupMetadata(val groupId: String, initialState: GroupState
+        } catch {
+          case e: SchemaException => {
+            warn(s"Failed to parse Consumer Protocol ${ConsumerProtocol.PROTOCOL_TYPE}:${protocol.get} " +
+              s"of group $groupId. Consumer Group is not aware of the subscribed topics.", e)
```
nit: Consumer group _coordinator_ is not aware?

hachikuji commented on core/src/main/scala/kafka/server/KafkaApis.scala:
```diff
@@ -2625,6 +2627,49 @@ class KafkaApis(val requestChannel: RequestChannel,
+    val offsetDeleteRequest = request.body[OffsetDeleteRequest]
+    val groupId = offsetDeleteRequest.data.groupId()
+
+    if (authorize(request.session, Read, Resource(Group, groupId, LITERAL))) {
```
I guess we didn't specify this in the KIP, but it seems reasonable to use the same permission as the OffsetCommit API. On the other hand, the DeleteGroups API requires Delete permission. Since they are both administrative use cases, do you think it makes sense for them to be consistent?

dajac replied:
Yeah. I thought it makes more sense to allow an application to delete an offset if it can commit it. On the other hand, we could argue that deleting is more an exceptional behaviour so it could make sense to have the possibility to restrict it. I think the later is more reasonable. I will update it.

hachikuji commented on core/src/main/scala/kafka/server/KafkaApis.scala:
```diff
@@ -2625,6 +2627,49 @@ class KafkaApis(val requestChannel: RequestChannel,
+    if (authorize(request.session, Read, Resource(Group, groupId, LITERAL))) {
+      val topicPartitions = offsetDeleteRequest.data.topics().asScala.flatMap { topic =>
+        topic.partitions().asScala.map { partition =>
+          new TopicPartition(topic.name(), partition.partitionIndex())
```
The OffsetCommit and OffsetFetch APIs check permissions on each topic included in the request. Should we do the same here? I think we're basically trying to avoid letting this API leak information about the topics in the cluster.

dajac replied:
We should do the same. I naively thought that all members of a group would all have the permissions to see the topic used in the group. It seems it is not true. I will add the check as you suggest.

dajac replied:
In the same time, I have added a check which verifies if the group knows the partition. `UNKNOWN_TOPIC_OR_PARTITION` is returned if the partition is unknown.

hachikuji commented on core/src/main/scala/kafka/coordinator/group/GroupMetadata.scala:
```diff
@@ -423,6 +434,51 @@ private[group] class GroupMetadata(val groupId: String, initialState: GroupState
+            members.map { case (_, member) =>
+              val buffer = ByteBuffer.wrap(member.metadata(protocol.get))
+              ConsumerProtocol.deserializeVersion(buffer)
+              ConsumerProtocol.deserializeSubscriptionV0(buffer).topics().asScala.toSet
```
It may be worth a comment explaining why we deserialize as version 0.

hachikuji commented on core/src/main/scala/kafka/server/KafkaApis.scala:
```diff
@@ -2625,6 +2627,49 @@ class KafkaApis(val requestChannel: RequestChannel,
+          offsetDeleteRequest.getErrorResponse(requestThrottleMs, groupError)
+        else {
+          val topics = new OffsetDeleteResponseData.OffsetDeleteResponseTopicCollection
+          topicErrors.groupBy(_._1.topic()).map { case (topic, topicPartitions) =>
```
nit: it's conventional in scala to leave off parenthesis for simple getters like `topic()`. There are a few similar cases below

hachikuji commented on core/src/main/scala/kafka/coordinator/group/GroupMetadata.scala:
```diff
@@ -204,6 +209,10 @@ private[group] class GroupMetadata(val groupId: String, initialState: GroupState
   private var receivedConsumerOffsetCommits = false
 
+  // When protocolType == `consumer`, a set of subscribed topics is maintained. The set is
+  // computed when a new generation is created or when the group is restored from the log.
```
So this means that offset deletion does not take into account the subscriptions of an active rebalance. That's probably fine I guess since it is simpler and we are not modifying the commit state for any partitions which are actively being fetched.

A couple questions for completeness:

1. For the sake of argument, suppose we loosened the validation so that we allow explicit deletions (but not expirations) without respect to the group or subscription state. The behavior of the consumer would be a little unpredictable. It might just write over the deletion. It might also force an offset reset if an OffsetFetch followed the deletion. Because the behavior is undefined, we do not allow it, which seems fair. However, the downside is that it makes the API a little harder to use, so just wanted to check whether you think it is a worthwhile restriction.

2. If we know the subscription, then it wouldn't be much trouble to delete the offsets of all partitions that are not currently subscribed to. This could be supported in the API by using a null in the OffsetDelete request. I think it could be useful, but I'm inclined to leave it for future work. What do you think?

dajac replied:
Yes and No. I have distinguished two cases when I designed this:

1) When the group is empty and the initial rebalance takes place, subscriptions will be ignored until the rebalance is completed and a new generation is initialised. I thought it would be inappropriate to compute the set of subscribed topics before having the full view of the consumer group. 

2) When the group is not empty and a rebalance takes place, the current set of subscribed topics will remain until all rebalance is completed and a new generation is initialised. I think this is a good think to be conservative here until the group has a full view of all the members.

Another aspect is that the protocol is actually selected at this point as well. I think it is better to stay consistent and parse the consumer protocol once the protocol of the group has been selected. It is not a hard requirement though as we could parse any versions but it feels more consistent in my opinion.

To answer your questions:

1) I think allowing explicit deletions without respect to the group or subscription state would be really risky as the behaviour of the consumer would be unpredictable as you pointed out. I would say that it does not make much sense anyway to delete offsets of a topic while a running consumer group consumes this topic. It is good to have the restriction in place in my opinion.

2) Yeah. It would be possible to delete the offsets of all partitions *of the topics* that are not currently subscribed to. I wonder it it really useful though as they will be eventually cleaned up now. I don't see the benefit of having the possibility to basically force the immediate clean up. I would leave it for future work and re-iterate based on the feedback we get from the users.

hachikuji replied:
Thanks for the response. Makes sense and I agree.

hachikuji commented on core/src/main/scala/kafka/coordinator/group/GroupCoordinator.scala:
```diff
@@ -519,6 +542,62 @@ class GroupCoordinator(val brokerId: Int,
     groupErrors
   }
 
+  def handleDeleteOffsets(groupId: String, partitions: Seq[TopicPartition]): (Errors, Map[TopicPartition, Errors]) = {
```
Leaving this here for lack of a better alternative. One thing that I didn't cover in the KIP is changes to the consumer group command to use this API. I think we can extend the KIP and do this in a separate PR. What do you think?

dajac replied:
It makes sense. Let's do it in a separate PR.

dajac replied:
I have opened a JIRA for it: https://issues.apache.org/jira/browse/KAFKA-8901

hachikuji reviewed (commented):
Thanks, the broker logic looks pretty good. I left a few comments/questions.

dajac committed:
Address reviews

dajac commented:
@hachikuji I think I have addressed all your questions and comments. Thanks for the review! I will rebase the PR to resolve conflicts once you are done with the review.

dajac committed:
Increase session timeout in integration test

dajac commented:
retest this please

dajac committed:
fix tests

dajac commented:
retest this please

hachikuji commented on clients/src/main/java/org/apache/kafka/common/requests/OffsetDeleteRequest.java:
```diff
@@ -0,0 +1,107 @@
+    public AbstractResponse getErrorResponse(int throttleTimeMs, Errors error) {
+        OffsetDeleteResponseTopicCollection topics = new OffsetDeleteResponseTopicCollection();
+
+        for (OffsetDeleteRequestTopic topic : data.topics()) {
```
In this case, would it be reasonable to set only the top level error? If the client sees a top-level error code which is not NONE, then there is no need to process partition-level errors.

dajac replied:
Totally. I have missed this one.

hachikuji commented on core/src/main/scala/kafka/coordinator/group/GroupMetadataManager.scala:
```diff
@@ -95,6 +96,26 @@ class GroupMetadataManager(brokerId: Int,
+      "group-coordinator-metrics",
+      "The rate of committed offsets"),
+    metrics.metricName("offset-commit-count",
+    "group-coordinator-metrics",
```
nit: alignment is a little inconsistent here

hachikuji reviewed (commented):
Just one more comment. This is looking good.

dajac committed:
Address reviews

dajac commented:
@hachikuji Thank you for the review. I have addressed the remaining points. The builds keep being killed after 240 minutes. Is it a hard limit set in Jenkins?

hachikuji commented:
@dajac I added another 30 minutes to the build time. Let's see if that's enough.

hachikuji reviewed (approved):
LGTM. Thanks for the PR!
